Catch errors when unmarshalling delta/doubleDelta encoded chunks

dmilstein opened the pull request:
This is (hopefully) a fix for #1653

Specifically, this makes it so that if the length for the stored delta/doubleDelta is somehow corrupted to be too small, the attempt to unmarshal will return an error.

The current (broken) behavior is to return a malformed chunk, which can then lead to a panic when there is an attempt to read header values.

The referenced issue proposed creating chunks with a minimum length -- I instead opted to just error on the attempt to unmarshal, since I'm not clear on how it could be safe to proceed when the length is incorrect/unknown.

The issue also talked about possibly "quarantining series", but I don't know the surrounding code well enough to understand how to make that happen.

(Also, in case you're wondering: I'm poking around in Prometheus as a way to brush up on my Go skills)

@beorn7

dmilstein commented:
Gack, just realized this broken another test (hence the CI builds failing).  I'll see if I can fix that problem.

fabxc commented:
Thanks. The change generally looks good to me. Maybe @beorn7 can help on the failing tests.

Generally, rather than adding a test function for each possible error, we try to do table-driven tests. This makes it easy to modify and extend as functionality changes and to add test scenarios as we discover bugs.

Here is an example: https://github.com/prometheus/alertmanager/blob/fabxc-meshsil/silence/silence_test.go#L501-L620

Sometimes it's most convenient to split this out into two functions, one testing the happy path and one testing all the different error conditions.

dmilstein commented:
First off, I figured out what was breaking the build -- turns out I didn't understand that doubleDeltaChunk can legally have a size less than the full header length.  I updated my code to no longer return an error in that case (and added a constant + comment in the doubledelta.go source).  Tests pass locally for me now (which they didn't before).

I also went ahead and took @fabxc's suggestions: inlined the helper func, and converted the tests to table-driven.  That mostly worked well, but I did run into an odd thing with the type signature of the constructor functions in the test fixtures.  I threw in a comment about that, if there's a way to make the compiler happy without the kludge-y bit, very happy to hear about it.

fabxc commented on storage/local/delta_test.go:
```diff
@@ -0,0 +1,106 @@
+			"deltaEncodedChunk",
+			// Hrm. Compiler didn't like a function which returned
+			// *deltaEncodedChunk for the chunkConstructor field. This is ugly,
+			// but I dunno how to make the compiler happy here.
```
Yup, those are entirely different type signatures to the compiler. The solution here is correct.

Please use the field names though in the struct initilizations. It's a more noisy but usually saves a lot of headache when the struct type is extended or its order modified.

fabxc commented:
@beorn7 for final check whether we didn't consider something relevant.

beorn7 commented:
Thanks for your contribution and the careful testing.

Merging once CI is green.
